### PR TITLE
Modular installer part3

### PR DESCRIPTION
**Pull request by IceReaper**
Part 1 #20440
Part 2 #20445 
Part 3 #20439

What we have here:
- The Blast decompression now properly supports the actual file format and does not need hardcoded offsets and lengths!
- Mix files are now properly extracted and do not need hardcoded offsets and lengths!
- Added support for the GoG and Steam Source Resolvers.
- Added installer support for C&C Remastered using Origin, Steam and the EA App.

### CODE REVIEW

File: OpenRA.Mods.Common/Installer/IExtractor.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.IO;
+
+namespace OpenRA.Mods.Common.Installer
+{
+	public interface IExtractor : IDisposable
```
Comment (pchote) [inline OpenRA.Mods.Common/Installer/IExtractor.cs]: This appears to mostly duplicate `IReadOnlyPackage`?

File: OpenRA.Mods.Common/Installer/SourceResolvers/RegistryDirectoryFromFileSourceResolver.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
+using System.IO;
```
Comment (pchote) [inline OpenRA.Mods.Common/Installer/SourceResolvers/RegistryDirectoryFromFileSourceResolver.cs]: Missing a file header.

File: OpenRA.Mods.Common/ModContent.cs [hunk 3]
```diff
@@ -41,7 +40,8 @@ public bool IsInstalled()
 
 		public class ModSource
 		{
-			public readonly SourceType Type = SourceType.Disc;
+			[FieldLoader.Ignore]
+			public readonly MiniYaml Type = new MiniYaml("Disc");
```
Comment (pchote) [inline OpenRA.Mods.Common/ModContent.cs]: Default to `null` and make `Type` required for all definitions?

File: OpenRA.Mods.Common/ModContent.cs [hunk 4]
```diff
@@ -78,6 +82,7 @@ public class ModDownload
 			public readonly string URL;
 			public readonly string MirrorList;
 			public readonly string SHA1;
+			public readonly string Extractor = "Zip";
```
Comment (pchote) [inline OpenRA.Mods.Common/ModContent.cs]: Likewise here.

File: OpenRA.Mods.Common/FileFormats/InstallShieldCABCompression.cs [hunk 5]
```diff
@@ -42,10 +42,14 @@ enum LinkFlags : byte
 			public readonly uint FirstFile;
 			public readonly uint LastFile;
 
-			public FileGroup(Stream stream, long offset)
+			public FileGroup(Stream stream, long offset, uint version)
```
Comment (pchote) [inline OpenRA.Mods.Common/FileFormats/InstallShieldCABCompression.cs]: These changes should be split to at least their own commit, maybe their own PR.

File: OpenRA.Mods.Common/FileExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+#endregion
+
+using System;
+using System.IO;
+using System.Linq;
+
+namespace OpenRA.Mods.Common
+{
+	public static class FileInsensitive
```
Comment (pchote) [inline OpenRA.Mods.Common/FileExtensions.cs]: I'm surprised that this needs to be implemented from scratch. Is there really not an existing solution?

**Review by pchote — COMMENTED** [review]
Partial review:

File: OpenRA.Mods.Common/FileExtensions.cs (hunk 6 above)
Comment (IceReaper) [reply OpenRA.Mods.Common/FileExtensions.cs]: I was also extremely surprised, when i found out that .net does not ship this by default.

Comment (pchote) [reply OpenRA.Mods.Common/FileExtensions.cs]: It would be good to have a comment here explaining what you mention in the PR description

**Comment by IceReaper** [comment]
I splitted the PR into several easier to review commits ;)

File: OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackageLogic.cs [hunk 7]
```diff
@@ -217,26 +218,36 @@ void OnDownloadProgress(long total, long read, int progressPercentage)
 						try
 						{
 							using (var stream = File.OpenRead(file))
-							using (var z = new ZipFile(stream))
 							{
-								foreach (var kv in download.Extract)
+								var packageLoaderName = $"{download.Type}Loader";
+
+								var packageLoaderType = AppDomain.CurrentDomain
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackageLogic.cs]: `var packageLoader = modData.ObjectCreator.CreateObject<IPackageLoader>($"{download.Type}Loader");` should be sufficient

File: OpenRA.Mods.Common/Installer/ISourceAction.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+#endregion
+
+using System;
+using System.Collections.Generic;
+
+namespace OpenRA.Mods.Common.Installer
+{
+	public interface ISourceAction
+	{
+		void Run(MiniYaml actionYaml, string path, ModData modData, List<string> extracted, Action<string> updateMessage);
```
Comment (pchote) [inline OpenRA.Mods.Common/Installer/ISourceAction.cs]: This name is too generic IMO. Can we call it something that better describes what it does?

File: OpenRA.Mods.Common/Installer/SourceActions/CopySourceAction.cs [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+namespace OpenRA.Mods.Common.Installer
+{
+	public class CopySourceAction : ISourceAction
+	{
+		public void Run(MiniYaml actionYaml, string path, ModData modData, List<string> extracted, Action<string> updateMessage)
+		{
+			var sourceDir = Path.Combine(path, actionYaml.Value);
+			foreach (var node in actionYaml.Nodes)
+			{
+				var sourcePath = Path.Combine(sourceDir, node.Value.Value).Replace("/./", "/");
```
Comment (pchote) [inline OpenRA.Mods.Common/Installer/SourceActions/CopySourceAction.cs]: Why is the `.Replace` needed? This seems like a bodge?

Also, what about paths relative to a named OpenRA directory?

File: OpenRA.Mods.Common/Installer/SourceActions/DeleteSourceAction.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+using System.IO;
+
+namespace OpenRA.Mods.Common.Installer
+{
+	public class DeleteSourceAction : ISourceAction
+	{
+		public void Run(MiniYaml actionYaml, string path, ModData modData, List<string> extracted, Action<string> updateMessage)
+		{
+			// Yaml path may be specified relative to a named directory (e.g. ^SupportDir) or the detected source path
+			var sourcePath = actionYaml.Value.StartsWith("^") ? Platform.ResolvePath(actionYaml.Value) : Path.Combine(path, actionYaml.Value);
```
Comment (pchote) [inline OpenRA.Mods.Common/Installer/SourceActions/DeleteSourceAction.cs]: What is the usecase for deleting files in the source path? The source isn't an OpenRA-owned directory, so we can't assume we will have permissions to do this!

File: OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromDiscLogic.cs [hunk 11]
```diff
@@ -263,82 +263,15 @@ void InstallFromDisc(string path, ModContent.ModSource modSource)
...
-									break;
-								}
-
-							default:
-								Log.Write("debug", $"Unknown installation command {i.Key} - ignoring");
-								break;
-						}
+						var sourceActionName = $"{i.Key}SourceAction";
+
+						var sourceActionType = AppDomain.CurrentDomain
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromDiscLogic.cs]: `var sourceAction = modData.ObjectCreator.CreateObject<ISourceAction>($"{i.Key}SourceAction");`

File: OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromSourceLogic.cs [hunk 12]
```diff
@@ -150,65 +149,46 @@ public InstallFromDiscLogic(Widget widget, ModData modData, ModContent content,
...
-				}
-
 				foreach (var kv in sources)
 				{
-					message = modData.Translation.GetString(SearchingDiscFor, Translation.Arguments("title", kv.Value.Title));
+					message = modData.Translation.GetString(SearchingSourceFor, Translation.Arguments("title", kv.Value.Title));
+
+					var sourceResolverName = $"{kv.Value.Type.Value}SourceResolver";
+
+					var sourceResolverType = AppDomain.CurrentDomain
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromSourceLogic.cs]: `var sourceResolver = modData.ObjectCreator.CreateObject<ISourceResolver>($"{kv.Value.Type.Value}SourceResolver");`

File: OpenRA.Mods.Common/FileFormats/Blast.cs [hunk 13]
```diff
@@ -80,6 +80,7 @@ public static void Decompress(Stream input, Stream output, Action<long, long> on
 
 			// log2(dictionary size) - 6
 			var dict = br.ReadBits(8);
+
```
Comment (pchote) [inline OpenRA.Mods.Common/FileFormats/Blast.cs]: Stray file modification?

File: OpenRA.Mods.Common/Installer/SourceResolvers/SteamSourceResolver.cs [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text.RegularExpressions;
+using Microsoft.Win32;
+
+namespace OpenRA.Mods.Common.Installer
+{
+	public class SteamSourceResolver : ISourceResolver
```
Comment (pchote) [inline OpenRA.Mods.Common/Installer/SourceResolvers/SteamSourceResolver.cs:22]: Is it too much scope creep to ask you to add this to TD and RA to copy the legacy assets out of the remaster's `Data/CNCDATA/RED_ALERT/CD1` (note: only has the allies disk, so we can't install soviet FMVs) and `Data/CNCDATA/TIBERIAN_DAWN/CD1`, `Data/CNCDATA/TIBERIAN_DAWN/CD2` respectively?

File: OpenRA.Mods.Common/Installer/SourceActions/DeleteSourceAction.cs (hunk 10 above)
Comment (IceReaper) [reply OpenRA.Mods.Common/Installer/SourceActions/DeleteSourceAction.cs]: Well... i didnt invent this :D It was already here. See: https://github.com/OpenRA/OpenRA/blob/bleed/mods/cnc/installer/firstdecade.yaml#L336
https://github.com/OpenRA/OpenRA/blob/bleed/mods/ra/installer/firstdecade.yaml#L32
https://github.com/OpenRA/OpenRA/blob/bleed/mods/ra/installer/firstdecade.yaml#L448
https://github.com/OpenRA/OpenRA/blob/bleed/mods/ts/installer/firstdecade.yaml#L58
https://github.com/OpenRA/OpenRA/blob/bleed/mods/ts/installer/firstdecade.yaml#L594

File: OpenRA.Mods.Common/Installer/SourceResolvers/SteamSourceResolver.cs (hunk 14 above)
Comment (IceReaper) [reply OpenRA.Mods.Common/Installer/SourceResolvers/SteamSourceResolver.cs:22]: No problem! Ill add that ;)

File: OpenRA.Mods.Common/Installer/SourceActions/DeleteSourceAction.cs (hunk 10 above)
Comment (pchote) [reply OpenRA.Mods.Common/Installer/SourceActions/DeleteSourceAction.cs]: I mean cases that don't start with the ^!

File: OpenRA.Mods.Common/Installer/SourceActions/CopySourceAction.cs (hunk 9 above)
Comment (IceReaper) [reply OpenRA.Mods.Common/Installer/SourceActions/CopySourceAction.cs]: I had a problem with getting a path which has /./ somewhere in the middle, as the copy yaml node has a value of "." which is somewhat put in the middle. Ill re-check wether this problem still persist and give some additional insights on that...

File: OpenRA.Mods.Common/Installer/SourceActions/DeleteSourceAction.cs (hunk 10 above)
Comment (IceReaper) [reply OpenRA.Mods.Common/Installer/SourceActions/DeleteSourceAction.cs]: Oh! Okay, ill add a change for this :)

File: mods/cnc/installer/steam.yaml [hunk 15]
```diff
@@ -0,0 +1,147 @@
+cncr-steam: C&C Remastered  Collection (Steam version, English)
```
Comment (pchote) [inline mods/cnc/installer/steam.yaml]: Double whitespace

File: mods/ra/installer/steam.yaml [hunk 16]
```diff
@@ -0,0 +1,193 @@
+cncr-steam: C&C Remastered  Collection (Steam version, English)
```
Comment (pchote) [inline mods/ra/installer/steam.yaml]: Here too.

File: mods/ra/installer/steam.yaml [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+			^SupportDir|Content/ra/v2/interior.mix: interior.mix
+			^SupportDir|Content/ra/v2/scores.mix: scores.mix
+			^SupportDir|Content/ra/v2/snow.mix: snow.mix
+			^SupportDir|Content/ra/v2/sounds.mix: sounds.mix
+			^SupportDir|Content/ra/v2/russian.mix: russian.mix
+			^SupportDir|Content/ra/v2/allies.mix: allies.mix
+			^SupportDir|Content/ra/v2/temperat.mix: temperat.mix
+			^SupportDir|Content/ra/v2/movies1.mix: movies1.mix
+		#ExtractMix: Data/CNCDATA/RED_ALERT/CD2/MAIN.MIX
+		#	^SupportDir|Content/ra/v2/movies2.mix: movies2.mix
```
Comment (pchote) [inline mods/ra/installer/steam.yaml]: Remove

File: mods/ra/installer/steam.yaml [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+			^SupportDir|Content/ra/v2/movies/snowbomb.vqa: snowbomb.vqa
+			^SupportDir|Content/ra/v2/movies/soviet1.vqa: soviet1.vqa
+			^SupportDir|Content/ra/v2/movies/sovtstar.vqa: sovtstar.vqa
+			^SupportDir|Content/ra/v2/movies/spy.vqa: spy.vqa
+			^SupportDir|Content/ra/v2/movies/tanya1.vqa: tanya1.vqa
+			^SupportDir|Content/ra/v2/movies/tanya2.vqa: tanya2.vqa
+			^SupportDir|Content/ra/v2/movies/toofar.vqa: toofar.vqa
+			^SupportDir|Content/ra/v2/movies/trinity.vqa: trinity.vqa
+		Delete: ^SupportDir|Content/ra/v2/movies1.mix
+		#ExtractMix: ^SupportDir|Content/ra/v2/movies2.mix
```
Comment (pchote) [inline mods/ra/installer/steam.yaml]: and these

**Comment by IceReaper** [comment]
So, what i need tested and sadly cannot test myself due to regional restrictions...
Everything has to be tested with the english releases of:
- ~~Origin, CnC Ultimate Collection, cnc~~
- ~~Origin, CnC Ultimate Collection, ra~~
- ~~Origin, CnC Ultimate Collection, ts~~
- ~~Origin, CnC Remastered, cnc~~
- ~~Origin, CnC Remastered, ra~~
- ~~EA Launcher, CnC Ultimate Collection, cnc~~
- ~~EA Launcher, CnC Ultimate Collection, ra~~
- ~~EA Launcher, CnC Ultimate Collection, ts~~
- ~~EA Launcher, CnC Remastered, cnc~~
- ~~EA Launcher, CnC Remastered, ra~~
- ~~Steam, CnC Remastered, cnc~~
- ~~Steam, CnC Remastered, ra~~

All verified working!

**Comment by pchote** [comment]
Screenshot 2022-11-09 at 20 16 27

There are a *lot* of duplicated entries in the content not found list.

Also confirmed that _Steam, CnC Remastered, ra_ works on macOS.

**Review by pchote — DISMISSED** [review]

**Comment by Mailaender** [comment]
Needs a rebase.

File: mods/d2k/installer/d2k.yaml [hunk 19]
```diff
@@ -0,0 +1,330 @@
+d2k: Dune 2000 (English)
+	Type: Disc
+	IDFiles:
+		MUSIC/AMBUSH.AUD: bd5926a56a83bc0e49f96037e1f909014ac0772a
```
Comment (Mailaender) [inline mods/d2k/installer/d2k.yaml:4]: Using https://archive.org/details/dune-2000 I run into a weird case sensitivity problem on Linux:

image

where the ISO is `MUSIC/AMBUSH.AUD` but the mounted filesystem is `music/ambush.aud` so the check fails.

**Review by Mailaender — CHANGES_REQUESTED** [review]

File: mods/d2k/installer/d2k.yaml (hunk 19 above)
Comment (IceReaper) [reply mods/d2k/installer/d2k.yaml:4]: Yup, thats what i addressed with the Case insensitive Part of part1?

Comment (IceReaper) [reply mods/d2k/installer/d2k.yaml:4]: Oh i might have skipped the id Files! I'll check

**Comment by Mailaender** [comment]
I think you can now remove the offset calculation utility command.

File: mods/d2k/installer/d2k.yaml (hunk 19 above)
Comment (IceReaper) [reply mods/d2k/installer/d2k.yaml:4]: I've rechecked: the ID files are also checked case insensitive

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/dab4d9edbdfdc09eff51f7a63762e693fb53976e...f984ef5df1581ff61db4cd6b2fd70166e7eec4b8)
